File: sdks/java/src/main/java/com/blok/nanoservice/testing/TestRunner.java
package com.blok.nanoservice.testing;

import com.blok.nanoservice.node.NodeHandler;
import com.blok.nanoservice.node.NodeRegistry;
import com.blok.nanoservice.types.*;

import java.util.Map;

/**
 * Test runner for executing nodes in-process without starting an HTTP server.
 * <p>
 * Example:
 * <pre>{@code
 * TestRunner runner = new TestRunner();
 * runner.register("hello", new HelloWorldNode());
 *
 * Context ctx = new MockContext().withBody(Map.of("name", "World")).build();
 * ExecutionResult result = runner.execute("hello", ctx, Map.of("prefix", "Hi"));
 *
 * assert result.isSuccess();
 * }</pre>
 */
public class TestRunner {

    private final NodeRegistry registry;

    public TestRunner() {
        this.registry = new NodeRegistry();
    }

    /**
     * Registers a node handler for testing.
     *
     * @param name    the node name
     * @param handler the node handler
     * @return this runner for fluent use
     */
    public TestRunner register(String name, NodeHandler handler) {
        registry.register(name, handler);
        return this;
    }

    /**
     * Executes a node by name with the given context and config.
     *
     * @param name   the node name
     * @param ctx    the execution context
     * @param config the node config map
     * @return the execution result
     */
    public ExecutionResult execute(String name, Context ctx, Map<String, Object> config) {
        NodeConfig nodeConfig = new NodeConfig();
        nodeConfig.setName(name);
        nodeConfig.setConfig(config);

        ExecutionRequest request = new ExecutionRequest(nodeConfig, ctx);
        return registry.execute(request);
    }
}

File: sdks/ruby/lib/nanoservice/testing/test_runner.rb
# frozen_string_literal: true

module Nanoservice
  module Testing
    # TestRunner provides in-process node execution for testing.
    # Register nodes, then execute them with mock contexts and configs.
    #
    # @example
    #   runner = Nanoservice::Testing::TestRunner.new
    #   runner.register("greet", GreetNode.new)
    #   result = runner.execute("greet", ctx, { "prefix" => "Hi" })
    #   assert result.success
    #
    class TestRunner
      def initialize
        @registry = Node::NodeRegistry.new("test")
      end

      # Register a node handler for testing.
      # @param name [String] Node name
      # @param handler [Node::NodeHandler] Handler instance
      # @return [self]
      def register(name, handler)
        @registry.register(name, handler)
        self
      end

      # Execute a node with a context and optional config.
      # @param name [String] Node name
      # @param ctx [Types::Context] Workflow context
      # @param config [Hash] Node configuration
      # @return [Types::ExecutionResult]
      def execute(name, ctx, config = {})
        node_config = Types::NodeConfig.new(name: name, config: config)
        request     = Types::ExecutionRequest.new(node: node_config, context: ctx)
        @registry.execute(request)
      end
    end
  end
end
